src/rust/collections/concurrent_ring.rs

```diff
@@ -0,0 +1,509 @@
+    fn do_enqueue_dequeue(ring: &mut ConcurrentRingBuffer) -> Result<()> {
+        // Insert items in the ring buffer.
+        let data: [u8; ITERATIONS] = [1; ITERATIONS];
+        let mut elements: usize = 0;
+        while ring.try_push(&data[0..elements]).is_ok() && elements < ITERATIONS {
+            elements = elements + 1;
+        }
+
+        // Check if buffer state is consistent.
+        //crate::ensure_eq!(ring.is_empty(), false);
+
+        let mut data_: [u8; ITERATIONS] = [0; ITERATIONS];
+        // Remove items from the ring buffer.
+        for i in 0..elements {
+            crate::ensure_eq!(ring.try_pop(&mut data_[0..i]).is_ok(), true);
+            crate::ensure_eq!(data_[elements], 1);
+        }
+
+        // Check if buffer state is consistent.
+        crate::ensure_eq!(ring.is_empty(), true);
+        crate::ensure_eq!(ring.is_full(), false);
+
+        Ok(())
+    }
+
+    /// Tests if we succeed to create a ring buffer with a valid capacity.
+    #[test]
+    fn new() -> Result<()> {
+        do_new()?;
+        Ok(())
+    }
+
+    /// Tests if we succeed to construct a ring buffer from raw parts.
+    #[test]
+    fn from_raw_parts() -> Result<()> {
+        const SIZE: usize = RING_BUFFER_CAPACITY + 2 * mem::size_of::<usize>();
+        let mut array: [u8; SIZE] = [0; SIZE];
+        do_from_raw(array.as_mut_ptr() as *mut u8, SIZE)?;
+        Ok(())
+    }
+
+    /// Tets if we succeed to sequentially enqueue and dequeue elements to/from a ring buffer.
+    #[test]
+    fn enqueue_dequeue_sequential() -> Result<()> {
+        let mut ring: ConcurrentRingBuffer = do_new()?;
+
+        do_enqueue_dequeue(&mut ring)
+    }
+
+    /// Tets if we succeed to sequentially enqueue and dequeue elements to/from a constructed ring buffer.
+    #[test]
+    fn enqueue_dequeue_sequential_raw() -> Result<()> {
+        const LENGTH: usize = RING_BUFFER_CAPACITY + 2 * mem::size_of::<usize>();
+        const SIZE: usize = LENGTH * mem::size_of::<u8>();
+        let mut array: [u8; LENGTH] = [0; LENGTH];
+        let mut ring: ConcurrentRingBuffer = do_from_raw(array.as_mut_ptr() as *mut u8, SIZE)?;
+
+        do_enqueue_dequeue(&mut ring)
+    }
+
+    /// Tests if we succeed to access a ring buffer concurrently.
+    #[test]
+    fn enqueue_dequeue_concurrent() -> Result<()> {
+        let ring: ConcurrentRingBuffer = do_new()?;
+        let mut result: Result<()> = Ok(());
+        thread::scope(|s| {
+            let mut writers: Vec<thread::ScopedJoinHandle<()>> = Vec::<thread::ScopedJoinHandle<()>>::new();
+            let mut readers: Vec<thread::ScopedJoinHandle<Result<()>>> =
+                Vec::<thread::ScopedJoinHandle<Result<()>>>::new();
+
+            for _ in 0..CONCURRENT_CLIENTS {
+                let writer: thread::ScopedJoinHandle<()> = s.spawn(|| {
+                    let data: [u8; ITERATIONS] = [1; ITERATIONS];
+                    let mut elements: usize = 0;
```

Comment:
Apparently all threads would add the same data in the buffers. May be a test that writes thread-specific data would be interesting to observe that there's no overwriting etc. happening...?

Not a blocker, we can do this later...

Reply:
Test has ben re-written.